File: go/0239H-滑动窗口最大值.go
package main

type MonotoneQueue struct {
	queue []int
}

func Constructor() MonotoneQueue {
	return MonotoneQueue{
		queue: make([]int, 0),
	}
}

func (this *MonotoneQueue) Front() int {
	return this.queue[0]
}

func (this *MonotoneQueue) Back() int {
	return this.queue[len(this.queue)-1]
}

func (this *MonotoneQueue) Empty() bool {
	return len(this.queue) == 0
}

func (this *MonotoneQueue) Push(x int) {
	for !this.Empty() && x > this.Back() {
		this.queue = this.queue[:len(this.queue)-1]
	}
	this.queue = append(this.queue, x)
}

func (this *MonotoneQueue) Pop(x int) {
	if !this.Empty() && this.Front() == x {
		this.queue = this.queue[1:]
	}
}

func maxSlidingWindow(nums []int, k int) []int {
	queue := Constructor()
	ans := make([]int, 0)
	for i := 0; i < k; i++ {
		queue.Push(nums[i])
	}
	ans = append(ans, queue.Front())
	for i := k; i < len(nums); i++ {
		queue.Pop(nums[i-k])
		queue.Push(nums[i])
		ans = append(ans, queue.Front())
	}
	return ans
}

File: cpp/0239H-滑动窗口最大值.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-20 14:28:04
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    deque<int> que;  // 单调队列，递减

    void pop(int val) {
        if (!que.empty() && val == que.front()) que.pop_front();
    }

    void push(int val) {
        while (!que.empty() && val > que.back()) que.pop_back();
        que.push_back(val);
    }

    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        for (int i = 0; i < k; i++) push(nums[i]);
        ans.push_back(que.front());
        for (int i = k; i < nums.size(); i++) {
            pop(nums[i - k]);
            push(nums[i]);
            ans.push_back(que.front());
        }
        return ans;
    }
};